examples/run/cuda/CMakeLists.txt

```diff
@@ -29,10 +29,10 @@ target_link_libraries( traccc_examples_cuda
    PUBLIC CUDA::cudart vecmem::core vecmem::cuda traccc::core
           traccc::device_common traccc::cuda )
 
-traccc_add_executable( throughput_st_cuda "throughput_st.cpp"
+traccc_add_executable( throughput_st_cuda "throughput_st.cpp" "inst_nvtx.cpp"
```

Comment:
We should aim to put this `inst_nvtx.cpp` file into a dedicated target which we can then link against using `target_link_libraries`. I'm not 100% sure how linkage works with `-finstrument-functions` but we should be able to compile this into a new library and then link. E.g.:

```cmake
traccc_add_library(nvtx_instrumentation "inst_nvtx.cpp")
```

And then...

```
traccc_add_executable( throughput_st_cuda
   LINK_LIBRARIES ... nvtx_instrumentation
)
```

Reply:
Should we put it in examples/options instead?

Reply:
No, just put it in this file. :)

Reply:
Not sure why but when i do:
`add_library(nvtx_instrumentation SHARED
"inst_nvtx.cpp")`

It is not able to find the nvtx.h file.

However, when I put it inside:
`add_library(traccc_examples_cuda)`

then it is able to find the header files
Do you know why this might be the case?

Reply:
So this is what I did:

`add_library(nvtx_instrumentation SHARED
        "inst_nvtx.cpp" )

target_link_libraries( nvtx_instrumentation
        PUBLIC CUDA::cudart traccc::core
        traccc::device_common traccc::cuda 


traccc_add_executable( throughput_st_cuda "throughput_st.cpp"
        LINK_LIBRARIES vecmem::core vecmem::cuda traccc::io traccc::performance
                  traccc::core traccc::device_common traccc::cuda
                  traccc::options traccc_examples_cuda
                  nvtx_instrumentation ${CMAKE_DL_LIBS})`

But is seems as though the instrumentation feature is not working if `inst_nvtx.cpp` is not added as an executable. Do you know why this might be?

Reply:
:thinking: I believe compilation-wise you'll need a setup like:

https://github.com/acts-project/vecmem/blob/main/sycl/CMakeLists.txt#L48-L77

Note that that code demonstrates multiple things:
  - There we also had to make sure that a piece of code would get linked **into** the client code, not just "against" it. A static library is a good way to go with that.
  - Static libraries by default are **not** built with position independent code. But since we needed PIC there as well, CMake was told to do so using [POSITION_INDEPENDENT_CODE](https://cmake.org/cmake/help/latest/prop_tgt/POSITION_INDEPENDENT_CODE.html). (This formalism makes sure that the correct flag would be used, even for compilers that don't use `-fPIC` for this.)

Reply:
Hi @krasznaa, so what I understand from the vecmem code you posted is that a check was firstly done on the source code to see if CUDA polyfill can make the test work, if it does then you add a library this file: `"src/utils/sycl/cuda_assert_polyfill.sycl"`. You then set the target properties so that the code uses position-independent code. I implemented all of this using `SHARED` libraries first, but this did not work. I then went to have a look at different types of libraries one can create and saw the option of generating `OBJECT` libraries. This seems to work. 

Here is the code I have so far:

`traccc_add_library(nvtx_instrumentation TYPE OBJECT
        "inst_nvtx.cpp" )`

`target_link_libraries( nvtx_instrumentation
        PUBLIC CUDA::cudart traccc::core
        traccc::device_common traccc::cuda )`

`traccc_add_executable( throughput_st_cuda "throughput_st.cpp"
        LINK_LIBRARIES vecmem::core vecmem::cuda traccc::io traccc::performance
                  traccc::core traccc::device_common traccc::cuda
                  traccc::options traccc_examples_cuda
                  nvtx_instrumentation ${CMAKE_DL_LIBS})`

Reply:
I hope this is fine..?

Reply:
Here is the link to the nvidia blog post you asked for: https://developer.nvidia.com/blog/cuda-pro-tip-generate-custom-application-profile-timelines-nvtx/